### PR TITLE
Allow checkpointing without flushing

### PR DESCRIPTION
**Pull request by siying**
Summary: Add a parameter to Checkpoint::CreateCheckpoint() so that flush can be skipped if total log file size is within a threshold.

Test Plan: Add unit test coverage in checkpoint_test

### CODE REVIEW

File: utilities/checkpoint/checkpoint_test.cc [hunk 1]
```diff
@@ -345,6 +347,63 @@ TEST_F(CheckpointTest, CheckpointCF) {
...
+  std::vector<ColumnFamilyHandle*> cphandles;
+
+  ASSERT_OK(DestroyDB(snapshot_name, options));
+  env_->DeleteDir(snapshot_name);
+
+  Status s;
+  // Take a snapshot
+  Checkpoint* checkpoint;
+  ASSERT_OK(Checkpoint::Create(db_, &checkpoint));
+  ASSERT_OK(checkpoint->CreateCheckpoint(snapshot_name, 1000000));
```
Comment (ajkr) [inline utilities/checkpoint/checkpoint_test.cc]: do you check the file is unflushed?

File: utilities/checkpoint/checkpoint_test.cc [hunk 2]
```diff
@@ -345,6 +347,63 @@ TEST_F(CheckpointTest, CheckpointCF) {
...
+  std::vector<ColumnFamilyDescriptor> column_families;
+  for (size_t i = 0; i < cfs.size(); ++i) {
+    column_families.push_back(ColumnFamilyDescriptor(cfs[i], options));
+  }
+  ASSERT_OK(DB::Open(options, snapshot_name, column_families, &cphandles,
+                     &snapshotDB));
+  ASSERT_OK(snapshotDB->Get(roptions, cphandles[0], "Default", &result));
+  ASSERT_EQ("Default", result);
+  ASSERT_OK(snapshotDB->Get(roptions, cphandles[1], "one", &result));
+  ASSERT_EQ("one", result);
```
Comment (ajkr) [inline utilities/checkpoint/checkpoint_test.cc]: weren't these added after the checkpoint was taken?

**Comment by ajkr** [comment]
Have you considered adding the flag `flush_before_checkpoint` similar to the `flush_before_backup` in the backup code path? Since I want to reunite these code paths it'd make my life easier :).

**Comment by siying** [comment]
@ajkr I don't want to add a new feature to backupDB just for code reuse purpose. We would need to support the feature if there is a bug or improvement request.

**Comment by siying** [comment]
@ajkr I'll work on the pending refactoring PR you have to make sure the two code path can be consolidated after this PR. I will don't think we should change backupDB interface because of it.

File: utilities/checkpoint/checkpoint_test.cc (hunk 2 above)
Comment (siying) [inline utilities/checkpoint/checkpoint_test.cc]: Before checkpoint:
```
  ASSERT_OK(Put(0, "Default", "Default"));
  ASSERT_OK(Put(1, "one", "one"));
  Flush();
  ASSERT_OK(Put(2, "two", "two"));
```
After checkpoint:
```
  ASSERT_OK(Put(1, "one", "two"));
  ASSERT_OK(Flush(1));
  ASSERT_OK(Put(2, "two", "twentytwo"));
```
The assert:
```
  ASSERT_OK(snapshotDB->Get(roptions, cphandles[0], "Default", &result));
  ASSERT_EQ("Default", result);
  ASSERT_OK(snapshotDB->Get(roptions, cphandles[1], "one", &result));
  ASSERT_EQ("one", result);
  ASSERT_OK(snapshotDB->Get(roptions, cphandles[2], "two", &result));
  ASSERT_EQ("two", result);
```

**Comment by siying** [comment]
Address other comments.

**Comment by siying** [comment]
@ajkr do you have more comments?

File: utilities/checkpoint/checkpoint_test.cc [hunk 3]
```diff
@@ -345,6 +347,71 @@ TEST_F(CheckpointTest, CheckpointCF) {
...
+  std::string result;
+  std::vector<ColumnFamilyHandle*> cphandles;
+
+  ASSERT_OK(DestroyDB(snapshot_name, options));
+  env_->DeleteDir(snapshot_name);
+
+  Status s;
+  // Take a snapshot
+  rocksdb::SyncPoint::GetInstance()->SetCallBack(
+      "DBImpl::BackgroundCallFlush:start", [&](void* arg) {
```
Comment (ajkr) [inline utilities/checkpoint/checkpoint_test.cc]: (y)

File: utilities/checkpoint/checkpoint_test.cc (hunk 2 above)
Comment (ajkr) [inline utilities/checkpoint/checkpoint_test.cc]: i missed the part before the checkpoint, thanks.

**Pull request closed without merging**